### PR TITLE
Hash map data structure utility

### PR DESCRIPTION
**Pull request by dylwhich**
### Description

Adds an implementation of everyone's favorite data structure! Includes support for string keys (the default), int-based keys, and includes helper functions to ease implementing custom hash functions for more complex types. It supports put, get, remove (and return), iteration, and remove-during-iteration operations. It dynamically resizes itself once it becomes too full.

### Test Instructions

Tested by copying the examples into a test mode and running them all.

### Readiness Checklist
- [x] I have run `make format` to format the changes
- [x] I have compiled the firmware and the changes have no warnings
- [x] I have compiled the emulator and the changes have no warnings
- [x] I have run `make cppcheck` and checked that `cppcheck_result.txt` has no warnings for the changes
- [x] I have added doxygen comments to any code used by more than one Swadge mode. This includes `/*! \file` comments with Design Philosophy, Usage, and Example sections for new headers.
- [x] I have run `make docs` and checked that `doxy_warnings.txt` has no warnings for the new code

### CODE REVIEW

**Comment by AEFeinstein** [comment]
Uhoh
```
main/utils/hashMap.c:114:34: error: implicit declaration of function 'reallocarray' [-Werror=implicit-function-declaration]
  114 |         hashBucket_t* newArray = reallocarray(map->values, size, sizeof(hashBucket_t));
      |                                  ^~~~~~~~~~~~
```

**Comment by dylwhich** [comment]
> Uh-oh, guess this doesn't exist for Windows GCC?
> 
> ```
> main/utils/hashMap.c:114:34: error: implicit declaration of function 'reallocarray' [-Werror=implicit-function-declaration]
>   114 |         hashBucket_t* newArray = reallocarray(map->values, size, sizeof(hashBucket_t));
>       |                                  ^~~~~~~~~~~~
> ```

Huh, apparently `reallocarray` is a BSD extension?

**Comment by AEFeinstein** [comment]
The more you know!

**Review by AEFeinstein — COMMENTED** [review]
Seems real nice overall. I'm a little surprised you went with multi-buckets rather than overwriting the prior value for a key collision, but it seems to work! I just want a little clarity on the `inline` stuff before merging.

File: main/utils/hashMap.c [hunk 1]
```diff
@@ -0,0 +1,1026 @@
...
+
+        // This bucket doesn't have multiple entries yet, so we need to convert it
+        if (!bucket->hasMulti)
+        {
+            // Gotta make it a multi-bucket
+            HASH_LOG("Bucket collision; converting bucket to multi-value");
+
+            // Copy the first node itself
+            newNode = malloc(sizeof(hashNode_t));
+            memcpy(newNode, node, sizeof(hashNode_t));
```
[the remark this answers is not in the document]
Comment (dylwhich) [reply main/utils/hashMap.c:305]: Ah so this is a result of the funky way I'm handling single-vs-multi-value buckets. If a bucket only has a single node, then the node it uses is one that's a member of the bucket. That node is union'd with the `list_t` used for multi-node buckets. So a single-node bucket doesn't need any dynamic allocation, but for a multi-node bucket it switches to having an individually allocated node as the value for each linked list node.

File: main/utils/hashMap.c [hunk 2]
```diff
@@ -0,0 +1,1026 @@
...
+
+    ///<  Whether the iterator has already been advanced, after removing the previous item
+    bool removed;
+} hashIterState_t;
+
+//==============================================================================
+// Static Function Prototypes
+//==============================================================================
+
+static inline void hashCheckSize(hashMap_t* map);
```
[the remark this answers is not in the document]
Comment (dylwhich) [reply main/utils/hashMap.c:87]: :man_shrugging: I probably just did it for More Performance but I can remove it. The compiler is probably ignoring it anyway...

File: main/utils/hashMap.h [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+ * The hash map also supports using structs or even primitives as key types, instead of strings.
+ * This can be done using hashInitBin() to supply custom hashing and comparison functions
+ * appropriate for the type of key being used.
+ *
+ * \subsubsection hashMap_example_nonStringKeys_enum Enum and Integer keys
+ *
+ * This example shows how to use a hash map where the keys are enum values. This method will
+ * work with any numeric type, as long as \c sizeof(type) is less than or equal to \c sizeof(void*) -- that
+ * is, as long as its value fits inside a pointer. In practice this means an \c int32_t can be used as a
+ * key in a hash map in this way, but an \c int64_t cannot.
```
[the remark this answers is not in the document]
Comment (dylwhich) [reply main/utils/hashMap.h:172]: Yeah this was also pretty annoying and I had to do some weird casting stuff to get rid of warnings when compiling to the emulator, which I'm hoping just get optimized away without having any negative effects on the ESP.

**Comment by dylwhich** [comment]
> Seems real nice overall. I'm a little surprised you went with multi-buckets rather than overwriting the prior value for a key collision, but it seems to work! I just want a little clarity on the `inline` stuff before merging.

To clarify the multi-bucket approach, that's just used when two items have the same value for `(hashCode % numberOfBuckets)` and end up in the same bucket -- if two values have the same actual key, the old value will still get overwritten like you'd expect.

File: main/utils/hashMap.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (AEFeinstein) [reply main/utils/hashMap.c]: I bet the compiler respects this actually. Back in the 8266 days I inlined (for speed) a function placed in IRAM (also for speed), and it inlined itself enough times to run out of IRAM and fail to link. That was fun to track down.

**Comment by AEFeinstein** [comment]
Oh, that makes way more sense. Don't try to do childcare and review code at the same time 😅

**Review by AEFeinstein — APPROVED** [review]

**Pull request merged**